modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/compute/ClientComputeTask.java

```diff
@@ -0,0 +1,166 @@
+import org.apache.ignite.internal.processors.platform.client.ClientCloseableResource;
+import org.apache.ignite.internal.processors.platform.client.ClientConnectionContext;
+import org.apache.ignite.internal.processors.platform.client.ClientNotification;
+import org.apache.ignite.internal.processors.platform.client.ClientObjectNotification;
+import org.apache.ignite.internal.processors.platform.client.ClientStatus;
+import org.apache.ignite.internal.processors.platform.client.IgniteClientException;
+import org.apache.ignite.internal.processors.task.GridTaskProcessor;
+import org.apache.ignite.internal.util.typedef.F;
+import org.apache.ignite.lang.IgnitePredicate;
+
+import static org.apache.ignite.internal.processors.platform.client.ClientMessageParser.OP_COMPUTE_TASK_FINISHED;
+import static org.apache.ignite.internal.processors.task.GridTaskThreadContextKey.TC_NO_FAILOVER;
+import static org.apache.ignite.internal.processors.task.GridTaskThreadContextKey.TC_NO_RESULT_CACHE;
+import static org.apache.ignite.internal.processors.task.GridTaskThreadContextKey.TC_SUBGRID_PREDICATE;
+import static org.apache.ignite.internal.processors.task.GridTaskThreadContextKey.TC_SUBJ_ID;
+import static org.apache.ignite.internal.processors.task.GridTaskThreadContextKey.TC_TIMEOUT;
+
+/**
+ * Client compute task.
+  */
+class ClientComputeTask implements ClientCloseableResource {
+    /** No failover flag mask. */
+    private static final byte NO_FAILOVER_FLAG_MASK = 0x01;
+
+    /** No result cache flag mask. */
+    private static final byte NO_RESULT_CACHE_FLAG_MASK = 0x02;
+
+    /** Context. */
+    private final ClientConnectionContext ctx;
+
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /** Task id. */
+    private volatile long taskId;
+
+    /** Task future. */
+    private volatile ComputeTaskInternalFuture<Object> taskFut;
+
+    /** Task closed flag. */
+    private final AtomicBoolean closed = new AtomicBoolean();
+
+    /**
+     * Ctor.
+     *
+     * @param ctx Connection context.
+     */
+    ClientComputeTask(ClientConnectionContext ctx) {
+        assert ctx != null;
+
+        this.ctx = ctx;
+
+        log = ctx.kernalContext().log(getClass());
+    }
+
+    /**
+     * @param taskId Task ID.
+     * @param taskName Task name.
+     * @param arg Task arguments.
+     * @param nodeIds Nodes to run task jobs.
+     * @param flags Flags for task.
+     * @param timeout Task timeout.
+     */
+    void execute(long taskId, String taskName, Object arg, Set<UUID> nodeIds, byte flags, long timeout) {
+        assert taskName != null;
+
+        this.taskId = taskId;
+
+        GridTaskProcessor task = ctx.kernalContext().task();
+
+        IgnitePredicate<ClusterNode> nodePredicate = F.isEmpty(nodeIds) ? F.alwaysTrue() : F.nodeForNodeIds(nodeIds);
+        UUID subjId = ctx.securityContext() == null ? null : ctx.securityContext().subject().id();
+
+        task.setThreadContextIfNotNull(TC_SUBGRID_PREDICATE, nodePredicate);
```

Comment:
nodePredicate looks like never can be null, so calling setThreadContext is preferable than setThreadContextIfNotNull